src/createActionListenerMiddleware.ts

```diff
@@ -0,0 +1,358 @@
+        type,
+        listener,
+        options
+      }
+    }
+  }
+) as BaseActionCreator<
+  {
+    type: string
+    listener: ActionListener<any, any, any, any>
+    options: ActionListenerOptions<any, any, any>
+  },
+  'actionListenerMiddleware/add'
+> & {
+  <
+    C extends TypedActionCreator<any>,
+    S,
+    D extends Dispatch,
+    O extends ActionListenerOptions<ReturnType<C>, S, D, When>
+  >(
+    actionCreator: C,
+    listener: ActionListener<ReturnType<C>, S, D, O>,
+    options?: O
+  ): AddListenerAction<ReturnType<C>, S, D, O>
+
+  <
+    S,
+    D extends Dispatch,
+    O extends ActionListenerOptions<AnyAction, S, D, When>
+  >(
+    type: string,
+    listener: ActionListener<AnyAction, S, D, O>,
+    options?: O
+  ): AddListenerAction<AnyAction, S, D, O>
+}
+
+interface RemoveListenerAction<
+  A extends AnyAction,
+  S,
+  D extends Dispatch<AnyAction>
+> {
+  type: 'actionListenerMiddleware/remove'
+  payload: {
+    type: string
+    listener: ActionListener<A, S, D, any>
+  }
+}
+
+/**
+ * @alpha
+ */
+export const removeListenerAction = createAction(
+  'actionListenerMiddleware/remove',
+  function prepare(
+    typeOrActionCreator: string | TypedActionCreator<string>,
+    listener: ActionListener<any, any, any, any>
+  ) {
+    const type =
+      typeof typeOrActionCreator === 'string'
+        ? typeOrActionCreator
+        : (typeOrActionCreator as TypedActionCreator<string>).type
+
+    return {
+      payload: {
+        type,
+        listener
+      }
+    }
+  }
+) as BaseActionCreator<
+  { type: string; listener: ActionListener<any, any, any, any> },
+  'actionListenerMiddleware/remove'
+> & {
+  <C extends TypedActionCreator<any>, S, D extends Dispatch>(
+    actionCreator: C,
+    listener: ActionListener<ReturnType<C>, S, D, any>
+  ): RemoveListenerAction<ReturnType<C>, S, D>
+
+  <S, D extends Dispatch>(
+    type: string,
+    listener: ActionListener<AnyAction, S, D, any>
+  ): RemoveListenerAction<AnyAction, S, D>
+}
+
+/**
+ * @alpha
+ */
+export function createActionListenerMiddleware<
+  S,
+  D extends Dispatch<AnyAction> = Dispatch
+>() {
+  type ListenerEntry = ActionListenerOptions<any, S, D, When> & {
+    listener: ActionListener<any, S, D, any>
+  }
+
+  const listenerMap: Record<string, Set<ListenerEntry> | undefined> = {}
+  const middleware: Middleware<
+    {
+      (action: Action<'actionListenerMiddleware/add'>): Unsubscribe
+    },
+    S,
+    D
+  > = api => next => action => {
+    if (addListenerAction.match(action)) {
+      const unsubscribe = addListener(
+        action.payload.type,
+        action.payload.listener,
+        action.payload.options
+      )
+
+      return unsubscribe
+    }
+    if (removeListenerAction.match(action)) {
+      removeListener(action.payload.type, action.payload.listener)
+
+      return
+    }
+
+    const listeners = listenerMap[action.type]
+    if (listeners) {
+      /* before */
+      for (const entry of listeners) {
```

Comment:
The logic for the "before" and "after" cases is _almost_ identical. Is it worth trying to deduplicate it?

Reply:
The duplicated code is pretty readable, but here's a stab at deduping it.

```TypeScript

  const defaultStage = 'before'
  if (listeners) {
    const triggerListener = (stage, listener) => {
      let stoppedPropagation = false

      if (
        (listener.when || defaultStage) === stage &&
        (!listener.condition || listener.condition(action, api.getState))
      ) {
        if (listener.once) {
          listeners.delete(listener)
        }

        listener.listener(action, {
          ...api,
          stopPropagation() {
            if (stage === 'after') {
              throw new Error(
                'stopPropagation can only be called by action listeners with the `when` option set to "before"'
              )
            }
            stoppedPropagation = true
          },
        })
      }

      return stoppedPropagation
    }

    /* before */
    for (const entry of listeners) {
      const stoppedPropagation = triggerListener('before', entry)
      if (stoppedPropagation) {
        return action
      }
    }

    const result = next(action)

    /* after */
    for (const entry of listeners) {
      triggerListener('before', entry)
    }

    return result
  }
  return next(action)

```

Reply:
those are now deduplicated & shrunk quite a bit